enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,441 @@
+
+```go
+type CertificateReference struct {
+	// certificate resource name.
+	// +optional
+	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+}
+
+Type RouteSpec struct {
+	// ...
+
+	CertificateRef CertificateReference `json:"certificateRef,omitempty" protobuf:"bytes,5,opt,name=certificateRef"`
+}
+```
+
+The reasoning for the field was mainly for future proofing the API to handle integrations
+with other third-party certificate management systems. But this API introduces inconsistency
+between how Ingresses are supported by cert-manager via `.spec.tls.secretName` and when using
+Routes, the user would have to provide the Certificate CR reference.
+
+This also adds coupling with external third-party API(s) and limits future enhancement to be
+a more generic solution for integrating with third-party certificate management systems.
+
+##### Alternative to new status condition types
+
+An alternative to introducing 3 new condition types, would be to use `metadata.managedFields`
+on Route to denote to the user which fields are managed by the secret-injector controller and
+server-side apply would be used by the controller during updates to ensure that `.spec.tls.certificate`
+and `.spec.tls.secret` are always managed by the controller as long as the annotation is present.
+
+This although helps in avoiding addition of new condition types will result in a few inconsistencies,
+
+- When transitioning from using managed certificates to manually managing custom certificates,
+  the `.managedFields` will still depict `.spec.tls` as managed by the controller since the update
+  operation done by the user is only on the `.metadata.annotations`.
+- Also since `oc` defaults to `--server-side=false`, mixing CSA and SSA results in ambiguous `.managedFields`.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The new controller (secret-injector) will be responsible for watching routes and processing those
+that have the valid annotation for `tls-secret-name`. This controller will also update
+`.status` of the route to contain the latest information on the certificate in use.
+
+Since secrets are a related resource to the secret-injector controller, the new controller
+will also set up watches and re-sync routes in the same namespace as that of the secret (similar
+to ingress-to-route controller).
+
+The new controller will watch all routes and process all routes in order to update
+`.status.certificate`. The ones with the annotation will have additional
+logic to update `.spec.tls.certificate` and `.spec.tls.key` based on the following pre-conditions,
+
+- Annotation is present and valid
+- Secret mentioned in the annotation exists and is valid
+
+In scenarios where the user deletes the annotation, it is the controller's
+responsibility to reset `.spec.tls` and this will be driven based on `.status` i.e.
+if the status denotes `ManagedCertificate` and since the required annotation is
```

Comment:
I suppose that "the latest certificate status condition has to be `ManagedCertificate`" would be the exact description of the controller's logic here, no?

Reply:
Addressed.